src/test/java/com/atlassian/db/replica/api/ThrottledCacheTest.java

```diff
@@ -0,0 +1,75 @@
+package com.atlassian.db.replica.api;
+
+import org.junit.Test;
+
+import java.util.ConcurrentModificationException;
+import java.util.Random;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatNoException;
+
+public class ThrottledCacheTest {
+    private final ExecutorService executor = Executors.newFixedThreadPool(10);
+
+    @Test
+    public void staleWhileInvalidating() throws InterruptedException {
+        ThrottledCache<Long> cache = new ThrottledCache<>();
+
+        cache.get(() -> 1L);
+        asyncPutWithSlowSupplier(cache, anyValue());
+
+        assertThat(cache.get(this::anyValue)).hasValue(1L);
+    }
+
+
+    @Test
+    public void serveLatestValue() {
+        ThrottledCache<Long> cache = new ThrottledCache<>();
+        cache.get(this::anyValue);
+        cache.get(this::anyValue);
+
+        assertThat(cache.get(() -> 4L)).hasValue(4L);
+    }
+
+    @Test
+    public void noConcurrentInvalidation() throws InterruptedException {
+        ThrottledCache<Long> cache = new ThrottledCache<>();
+
+        asyncPutWithSlowSupplier(cache, anyValue());
+        cache.get(() -> {
+            throw new ConcurrentModificationException("Concurrent invalidation is not allowed");
+        });
+
+        assertThatNoException();
+    }
+
+    @Test
+    public void emptyWhenLoadingFirstTime() throws InterruptedException {
+        ThrottledCache<Long> cache = new ThrottledCache<>();
+
+        asyncPutWithSlowSupplier(cache, anyValue());
+
+        assertThat(cache.get(this::anyValue)).isEmpty();
+    }
+
+
+    private void asyncPutWithSlowSupplier(ThrottledCache<Long> cache, long value) throws InterruptedException {
+        executor.submit(() -> {
+            cache.get(() -> {
+                try {
+                    Thread.sleep(1000);
```

Comment:
I've changed it to 1000  -> 100, and below  100-> 10 and all tests passed.

Reply:
Did you re-run 1k times? It may appear flaky on slower machines.

Reply:
My point is that you might use countdown latch to test it deterministically. thought its okish

Reply:
I changed the 100ms as it's affecting execution time. 1s is async time - we are not waiting for it.